app/src/main/java/org/simple/clinic/patient/PatientRepository.kt

```diff
@@ -56,6 +56,20 @@ class PatientRepository @Inject constructor(
 
   private var ongoingNewPatientEntry: OngoingNewPatientEntry = OngoingNewPatientEntry()
 
+  fun search(criteria: PatientSearchCriteria): Observable<List<PatientSearchResult>> {
+    return when(criteria) {
+      is PatientSearchCriteria.ByName -> search(criteria.patientName)
+      is PatientSearchCriteria.ByPhoneNumber -> search(criteria.phoneNumber)
```

Comment:
`PatientSearchCriteria` only communicates what type of search should be performed on a particular `search string`. That is also synonymous with the way our search looks. We decide criteria based on input.

Passing the actual `search string` to `PatientSearchCriteria` types doesn't encapsulate information about that sealed class type itself. The `search string` can exist outside the criteria so as to make it more generic.

`search(searchString, criteria)`

Another way to do this could be to enforce to implement a search mechanism for each `PatientSearchCriteria` using an `abstract` method.
Wdyt?

Reply:
I also realised that if we use the second technique we can eliminate the need to add `when` blocks everywhere.

Reply:
The problem with moving the search string out of the search criteria is that future types of search criteria which might need more information than just a string will no longer be supported. For example, if there's a need to search with a name and colony, then we can just change the `ByName` criteria to `ByNameAndColony` without affecting anything else. It is just an implementation detail at the moment that both the different search criteria take in only a string as a parameter, which is evident from the different names for the parameters.

You do have a great point about having different classes which implement different search mechanisms based on the criteria, though. I will do this in a different PR because I had to temporarily disable performance time tracking for the search by name flow which I need to fix first.

Reply:
> For example, if there's a need to search with a name and colony, then we can just change the ByName criteria to ByNameAndColony without affecting anything else.

Makes sense.